crates/stages/src/trie/mod.rs

```diff
@@ -0,0 +1,584 @@
+    tx: &'tx Transaction<'itx, DB>,
+}
+
+// TODO: implement caching and bulk inserting
+impl<'tx, 'itx, DB> cita_trie::DB for HashDatabase<'tx, 'itx, DB>
+where
+    DB: Database,
+{
+    type Error = TrieError;
+
+    fn get(&self, key: &[u8]) -> Result<Option<Vec<u8>>, Self::Error> {
+        Ok(self.tx.get::<tables::AccountsTrie>(H256::from_slice(key))?)
+    }
+
+    fn contains(&self, key: &[u8]) -> Result<bool, Self::Error> {
+        Ok(<Self as cita_trie::DB>::get(self, key)?.is_some())
+    }
+
+    fn insert(&self, key: Vec<u8>, value: Vec<u8>) -> Result<(), Self::Error> {
+        self.tx.put::<tables::AccountsTrie>(H256::from_slice(key.as_slice()), value)?;
+        Ok(())
+    }
+
+    fn remove(&self, key: &[u8]) -> Result<(), Self::Error> {
+        self.tx.delete::<tables::AccountsTrie>(H256::from_slice(key), None)?;
+        Ok(())
+    }
+
+    fn flush(&self) -> Result<(), Self::Error> {
+        Ok(())
+    }
+}
+
+impl<'tx, 'itx, DB: Database> HashDatabase<'tx, 'itx, DB> {
+    fn new(tx: &'tx Transaction<'itx, DB>) -> Result<Self, TrieError> {
+        let root = EMPTY_ROOT;
+        if tx.get::<tables::AccountsTrie>(root)?.is_none() {
+            tx.put::<tables::AccountsTrie>(root, [EMPTY_STRING_CODE].to_vec())?;
+        }
+        Ok(Self { tx })
+    }
+
+    fn new_with_root(tx: &'tx Transaction<'itx, DB>, root: H256) -> Result<Self, TrieError> {
+        if root == EMPTY_ROOT {
+            return Self::new(tx)
+        }
+        tx.get::<tables::AccountsTrie>(root)?.ok_or(TrieError::MissingRoot(root))?;
+        Ok(Self { tx })
+    }
+}
+
+/// Database wrapper implementing HashDB trait.
+struct DupHashDatabase<'tx, 'itx, DB: Database> {
+    tx: &'tx Transaction<'itx, DB>,
+    key: H256,
+}
+
+// TODO: implement caching and bulk inserting
+impl<'tx, 'itx, DB> cita_trie::DB for DupHashDatabase<'tx, 'itx, DB>
+where
+    DB: Database,
+{
+    type Error = TrieError;
+
+    fn get(&self, key: &[u8]) -> Result<Option<Vec<u8>>, Self::Error> {
+        let mut cursor = self.tx.cursor_dup_read::<tables::StoragesTrie>()?;
+        Ok(cursor.seek_by_key_subkey(self.key, H256::from_slice(key))?.map(|entry| entry.node))
+    }
+
+    fn contains(&self, key: &[u8]) -> Result<bool, Self::Error> {
+        Ok(<Self as cita_trie::DB>::get(self, key)?.is_some())
+    }
+
+    fn insert(&self, key: Vec<u8>, value: Vec<u8>) -> Result<(), Self::Error> {
+        self.tx.put::<tables::StoragesTrie>(
+            self.key,
+            StorageTrieEntry { hash: H256::from_slice(key.as_slice()), node: value },
+        )?;
+        Ok(())
+    }
+
+    fn remove(&self, key: &[u8]) -> Result<(), Self::Error> {
+        let opt_value = <Self as cita_trie::DB>::get(self, key)?;
+        if let Some(value) = opt_value {
+            self.tx.delete::<tables::StoragesTrie>(
+                self.key,
+                Some(StorageTrieEntry { hash: H256::from_slice(key), node: value }),
+            )?;
+        }
```

Comment:
nit: It is more performant to use the cursor to `seek_by_key_subkey` and used `delete_current` this would search the values only once.

Reply:
Done